crates/ty_python_semantic/src/dunder_all.rs

```diff
@@ -0,0 +1,469 @@
+    salsa::CycleRecoveryAction::Iterate
+}
+
+fn dunder_all_names_cycle_initial(_db: &dyn Db, _file: File) -> Option<FxHashSet<Name>> {
+    None
+}
+
+/// Returns a set of names in the `__all__` variable for `file`, [`None`] if it is not defined or
+/// if it contains invalid elements.
+pub(crate) fn dunder_all_names(db: &dyn Db, file: File) -> Option<&FxHashSet<Name>> {
+    #[allow(clippy::ref_option)]
+    #[salsa::tracked(return_ref, cycle_fn=dunder_all_names_cycle_recover, cycle_initial=dunder_all_names_cycle_initial)]
+    fn dunder_all_names_impl(db: &dyn Db, file: File) -> Option<FxHashSet<Name>> {
+        let _span = tracing::trace_span!("dunder_all_names", file=?file.path(db)).entered();
+
+        let module = parsed_module(db.upcast(), file);
+        let index = semantic_index(db, file);
+        let mut collector = DunderAllNamesCollector::new(db, file, index);
+        collector.visit_body(module.suite());
+        collector.into_names()
+    }
+
+    dunder_all_names_impl(db, file).as_ref()
+}
+
+/// A visitor that collects the names in the `__all__` variable of a module.
+struct DunderAllNamesCollector<'db> {
+    db: &'db dyn Db,
+    file: File,
+
+    /// The scope in which the `__all__` names are being collected from.
+    ///
+    /// This is always going to be the global scope of the module.
+    scope: ScopeId<'db>,
+
+    /// The semantic index for the module.
+    index: &'db SemanticIndex<'db>,
+
+    /// The origin of the `__all__` variable in the current module, [`None`] if it is not defined.
+    origin: Option<DunderAllOrigin>,
+
+    /// A flag indicating whether the module uses unrecognized `__all__` idioms or there are any
+    /// invalid elements in `__all__`.
+    invalid: bool,
+
+    /// A set of names found in `__all__` for the current module.
+    names: FxHashSet<Name>,
+}
+
+impl<'db> DunderAllNamesCollector<'db> {
+    fn new(db: &'db dyn Db, file: File, index: &'db SemanticIndex<'db>) -> Self {
+        Self {
+            db,
+            file,
+            scope: global_scope(db, file),
+            index,
+            origin: None,
+            invalid: false,
+            names: FxHashSet::default(),
+        }
+    }
+
+    /// Updates the origin of `__all__` in the current module.
+    ///
+    /// This will clear existing names if the origin is changed to mimic the behavior of overriding
+    /// `__all__` in the current module.
+    fn update_origin(&mut self, origin: DunderAllOrigin) {
+        if self.origin.is_some() {
+            self.names.clear();
+        }
+        self.origin = Some(origin);
+    }
+
+    /// Extends the current set of names with the names from the given expression which can be
+    /// either a list of names or a submodule's `__all__` variable.
+    ///
+    /// Returns `true` if the expression is a valid list or submodule `__all__`, `false` otherwise.
+    fn extend_from_list_or_submodule(&mut self, expr: &ast::Expr) -> bool {
```

Comment:
weak preference for not using "submodule" in all these wordings, even though the spec does, since I don't think it's actually required to be a submodule